File: sdk/ios/Sources/TraceletSDK/SmartMotionCoordinator.swift
import Foundation
import CoreLocation

/// Coordinates between MotionDetector (Accelerometer/ActivityRecognition) and
/// SpeedMotionManager (GPS Speed) to implement the 'smart' MotionDetectionMode.
///
/// Implements a Logical OR:
/// If EITHER sensor detects motion, the app stays in continuous tracking mode.
/// If BOTH sensors detect stationary, the app stops continuous tracking and relies
/// on Geofences/periodic wakeups to save battery.
public class TraceletSmartMotionCoordinator {
    
    private var coreCoordinator: SmartMotionCoordinator?
    
    public var isAccelMoving: Bool { coreCoordinator?.isAccelMoving() ?? false }
    public var isSpeedMoving: Bool { coreCoordinator?.isSpeedMoving() ?? true }
    
    private weak var sdk: TraceletSdk?
    
    public init(sdk: TraceletSdk) {
        self.sdk = sdk
        
        let useGeofences = sdk.configManager?.getStationaryTrackingMode() == .geofences
        self.coreCoordinator = SmartMotionCoordinator(useGeofencesWhenStationary: useGeofences)
    }
    
    /// Synchronize the Rust core mode with the native StateManager on startup or mode change.
    ///
    /// `stateManager.trackingMode` and the coordinator's `currentMode` measure
    /// different things, and conflating them wedged the coordinator (#344):
    ///
    /// * `trackingMode` is the **session** mode — which start API was called. It
    ///   is set to `.continuous` by `start()` and stays there for the whole
    ///   session; `switchToStationaryPeriodicForce()` deliberately leaves it
    ///   alone and records the pace in `stateManager.isMoving` instead.
    /// * `currentMode` is the **posture** — whether the engine is running
    ///   continuous GPS right now or is parked in a stationary schedule.
    ///
    /// `start()` sets `isMoving` from `motion.isMoving` (false by default) and
    /// then calls this, so mapping `.continuous -> .continuous` wrote
    /// `Continuous` into a coordinator whose inputs both said stationary. The
    /// core's `evaluate_state` emits no action for that pair in either
    /// direction: a moving accelerometer sees `currentMode == .continuous` and
    /// returns `.none`, so `switchToContinuousForce()` never ran and the session
    /// stayed parked — no fixes recorded, nothing to sync — until the process
    /// was killed. Reading the posture off `isMoving` keeps the two in step, and
    /// the first real accel or speed event produces the wake-up.
    ///
    /// #409 is the same wedge from the other side: `isMoving` false with a *live*
    /// stream parked the coordinator while the engine kept streaming, and
    /// `evaluate_state` returns `.none` for that pair too. The posture is now the
    /// OR of the committed pace and the engine's actual state — the only reading
    /// that survives both directions.
    ///
    /// The engine's state is `isContinuousStreaming`, not `isTracking`: the
    /// speed/smart pipeline parks by stopping continuous updates while leaving
    /// `isTracking` true so delegate callbacks keep arriving, so `isTracking`
    /// reads `true` on a parked engine. ORing *that* in writes Continuous into a
    /// coordinator that is genuinely parked, and the core emits no wake-up for a
    /// posture it already believes is Continuous — #344's swallowed shake,
    /// re-entered through the #409 fix.
    public func syncCurrentMode() {
        guard let stateManager = sdk?.stateManager else { return }

        let useGeofences = sdk?.configManager?.getStationaryTrackingMode() == .geofences
        coreCoordinator?.setCurrentMode(
            mode: TraceletSmartMotionCoordinator.coordinatorMode(
                sessionMode: stateManager.trackingMode,
                isMoving: stateManager.isMoving,
                isStreamLive: sdk?.locationEngine?.isContinuousStreaming ?? false,
                useGeofencesWhenStationary: useGeofences))
        coreCoordinator?.setUseGeofencesWhenStationary(useGeofences: useGeofences)
    }

    /// The posture the coordinator should be holding, given the session mode, the
    /// committed pace, and whether the engine is streaming right now. Pure so the
    /// mappings in #344 and #409 can be pinned directly.
    static func coordinatorMode(
        sessionMode: TraceletTrackingMode,
        isMoving: Bool,
        isStreamLive: Bool,
        useGeofencesWhenStationary: Bool
    ) -> TrackingMode {
        let stationary: TrackingMode = useGeofencesWhenStationary ? .stationaryGeofences : .stationaryPeriodic
        switch sessionMode {
        case .continuous:
            // `isMoving` predicts what start() is *about* to do; `isStreamLive`
            // reports what the engine is doing *now*. Either alone gets a case
            // wrong, so the posture is the OR of them (#409).
            //
            // `isMoving` alone was the wedge. start() syncs before it touches the
            // engine, so a session resuming stationary while a stream was still
            // live wrote a stationary posture into a coordinator whose engine was
            // streaming — and `evaluate_state` only emits the stop action when it
            // believes the posture is `.continuous`. Every later stationary
            // decision returned `.none` and GPS ran at the configured interval for
            // the rest of the session, on a device correctly reported as parked.
            //
            // `isStreamLive` alone breaks #344's direction: this runs before
            // start() starts the engine, so a moving start would sync stationary
            // and then stream — the same wedge, mirrored.
            return (isMoving || isStreamLive) ? .continuous : stationary
        case .geofences:
            return .stationaryGeofences
        case .periodic:
            return .stationaryPeriodic
        }
    }
    
    /// Called when the accelerometer/activity recognition state changes.
    /// Returns the action taken so the caller can conditionally reset
    /// the speed state machine on real wake-ups.
    @discardableResult
    public func onAccelStateChange(isMoving: Bool) -> CoordinatorAction {
        guard let action = coreCoordinator?.onAccelStateChange(isMoving: isMoving) else { return .none }
        TraceletLog.debug("[Tracelet] SmartMotionCoordinator: onAccelStateChange -> isMoving=\(isMoving), action=\(action)")
        handleAction(action)
        return action
    }
    
    /// GPS speed at or below which a *moving* accelerometer is read as hand
    /// tremor on a physically still device rather than as real motion.
    private static let tremorSpeedThreshold: Double = 0.15

    /// The last speed this process actually resolved, or `nil` if it has not
    /// resolved one yet.
    ///
    /// `lastEffectiveSpeed` and `lastLocation` are written together on every
    /// accepted fix, so a nil location is exactly "no fix has been accepted in
    /// this process" — which is *unknown*, not zero.
    private var resolvedSpeed: Double? {
        guard let engine = sdk?.locationEngine, engine.getLastLocation() != nil else { return nil }
        return engine.lastEffectiveSpeed
    }

    /// Returns the action taken, mirroring ``onAccelStateChange(isMoving:)``.
    /// Discardable — every existing caller ignores it — but it is the only
    /// externally visible product of this call, so a test has nothing else to
    /// assert the speed wake-up on.
    @discardableResult
    public func onSpeedStateChange(isMoving: Bool) -> CoordinatorAction {
        var overrideAction: CoordinatorAction = .none
        if !isMoving && isAccelMoving {
            // The speed machine says stationary and the accelerometer disagrees.
            // Two situations look like this:
            //   a) the device is physically still and the accelerometer is
            //      picking up hand tremor — GPS speed is genuinely ~0;
            //   b) the user is walking slowly, below the speed threshold — the
            //      accelerometer is the one telling the truth.
            // Only (a) justifies overruling a positive motion signal.
            //
            // #333: this used to read `getLastLocation()?.speed`, the raw
            // CLLocation value, which is **-1** on a fix that carries no speed —
            // and -1 sails through a `<= 0.15` test. An unknown speed was
            // therefore treated as proof the device was parked, and the
            // accelerometer got overridden on missing data. `lastEffectiveSpeed`
            // is the resolved value (documented on its declaration as existing
            // precisely because the cached CLLocation.speed "may be stale, 0, or
            // -1"), and no resolved speed at all now leaves the accelerometer
            // standing rather than silently siding against it.
            switch resolvedSpeed {
            case .none:
                TraceletLog.debug("[Tracelet] SmartMotionCoordinator: no GPS speed resolved yet — trusting accel, staying continuous.")
            case .some(let speed) where speed <= TraceletSmartMotionCoordinator.tremorSpeedThreshold:
                TraceletLog.debug(String(format: "[Tracelet] SmartMotionCoordinator: GPS speed near zero (%.2f m/s) — overriding accel to false (hand tremor).", speed))
                // Through the wrapper, so the action this produces is *handled*.
                // It used to be `_ = coreCoordinator?.onAccelStateChange(false)`,
                // and the accel flag is the second half of the OR: when the speed
                // flag is already stationary, flipping it here is the transition
                // that carries the whole stop decision. `evaluate_state` returned
                // `.switchToStationaryPeriodic` and it went on the floor, then the
                // `onSpeedStateChange` below deduped to `.none` because the flag
                // it would have flipped was already false — so nothing was ever
                // handled, the engine was never told, and continuous GPS ran on a
                // parked device for the rest of the session (#409).
                overrideAction = onAccelStateChange(isMoving: false)
            case .some(let speed):
                TraceletLog.debug(String(format: "[Tracelet] SmartMotionCoordinator: GPS speed %.2f m/s is above the %.2f m/s tremor threshold — trusting accel, staying continuous.",
                                         speed, TraceletSmartMotionCoordinator.tremorSpeedThreshold))
            }
        }
        guard let action = coreCoordinator?.onSpeedStateChange(isMoving: isMoving) else { return overrideAction }
        TraceletLog.debug("[Tracelet] SmartMotionCoordinator: onSpeedStateChange -> isMoving=\(isMoving), action=\(action)")
        handleAction(action)
        // The override above already switched the engine when the speed flag was
        // the one that was stale; reporting `.none` here would hide that from a
        // caller that reads the result.
        return action == .none ? overrideAction : action
    }
    
    /// Called when the user manually forces the pace via changePace().
    public func onManualPaceChange(isMoving: Bool) {
        if let accelAction = coreCoordinator?.onAccelStateChange(isMoving: isMoving) {
            handleAction(accelAction)
        }
        if let speedAction = coreCoordinator?.onSpeedStateChange(isMoving: isMoving) {
            handleAction(speedAction)
        }
    }
    
    private func handleAction(_ action: CoordinatorAction) {
        guard let sdk = sdk else { return }

        switch action {
        case .switchToContinuous:
            recordModeSwitch("CONTINUOUS")
            sdk.switchToContinuousForce()
            sdk.motionDetector.onManualPaceChange(true)

        case .switchToStationaryGeofences:
            recordModeSwitch("STATIONARY_GEOFENCES")
            sdk.switchToStationaryGeofencesForce()
            sdk.motionDetector.onManualPaceChange(false)

        case .switchToStationaryPeriodic:
            recordModeSwitch("STATIONARY_PERIODIC")
            sdk.switchToStationaryPeriodicForce()
            sdk.motionDetector.onManualPaceChange(false)

        case .none:
            break
        }
    }

    /// Records a tracking-mode switch on the always-on lifecycle channel,
    /// naming both inputs of the OR (#334).
    ///
    /// The coordinator is a logical OR of the accelerometer and the GPS-speed
    /// machine, so a downgrade means *both* said stationary. Which one flipped
    /// last — and what the other was doing at the time — is the first question
    /// asked of any "it stopped tracking while I was moving" report, and at the
    /// shipped log levels the answer was not recorded anywhere.
    private func recordModeSwitch(_ mode: String) {
        let speedText = resolvedSpeed.map { String(format: "%.2f", $0) } ?? "unknown"
        TraceletLog.lifecycle(
            "smart-motion: switching to \(mode) — accelMoving=\(isAccelMoving) "
                + "speedMoving=\(isSpeedMoving) lastSpeed=\(speedText)m/s")
    }
}

File: sdk/android/tracelet-sdk/src/main/kotlin/com/ikolvi/tracelet/sdk/motion/SmartMotionCoordinator.kt
package com.ikolvi.tracelet.sdk.motion

import android.content.Context
import com.ikolvi.tracelet.sdk.ConfigManager
import com.ikolvi.tracelet.sdk.TraceletEventSender
import com.ikolvi.tracelet.sdk.location.LocationEngine
import com.ikolvi.tracelet.sdk.service.LocationService
import com.ikolvi.tracelet.sdk.location.PeriodicLocationWorker
import com.ikolvi.tracelet.sdk.model.TrackingMode
import com.ikolvi.tracelet.sdk.StateManager
import com.ikolvi.tracelet.sdk.util.TraceletLogger

/**
 * Coordinates between MotionDetector (Accelerometer/ActivityRecognition) and
 * SpeedMotionManager (GPS Speed) to implement the 'smart' MotionDetectionMode.
 *
 * Implements a Logical OR:
 * If EITHER sensor detects motion, the app stays in continuous tracking mode.
 * If BOTH sensors detect stationary, the app stops continuous tracking and relies
 * on Geofences/periodic wakeups to save battery.
 */
class SmartMotionCoordinator(
    private val context: Context,
    private val configManager: ConfigManager,
    private val stateManager: StateManager,
    var events: TraceletEventSender,
    private val locationEngine: LocationEngine,
    private val motionDetector: MotionDetector,
    private val logger: TraceletLogger,
) {
    private val coreCoordinator = uniffi.tracelet_core.SmartMotionCoordinator(
        configManager.getStationaryTrackingMode() == com.ikolvi.tracelet.sdk.model.StationaryTrackingMode.GEOFENCES
    )

    val isAccelMoving: Boolean
        get() = coreCoordinator.isAccelMoving()

    val isSpeedMoving: Boolean
        get() = coreCoordinator.isSpeedMoving()

    /**
     * Called when the accelerometer/activity recognition state changes.
     * Returns the action taken by the coordinator so the caller can
     * conditionally reset the speed state machine on real wake-ups.
     */
    fun onAccelStateChange(isMoving: Boolean): uniffi.tracelet_core.CoordinatorAction {
        val action = coreCoordinator.onAccelStateChange(isMoving)
        logger.debug("SmartMotionCoordinator: onAccelStateChange -> isMoving=$isMoving, action=$action, isAccelMoving=${coreCoordinator.isAccelMoving()}, isSpeedMoving=${coreCoordinator.isSpeedMoving()}")
        handleAction(action)
        return action
    }

    /**
     * The last speed this process actually resolved, or `null` if it has not
     * resolved one yet.
     *
     * [LocationEngine.lastEffectiveSpeed] and `lastLocation` are written
     * together on every accepted fix, so a null location is exactly "no fix has
     * been accepted in this process" — which is *unknown*, not zero.
     */
    private val resolvedSpeed: Double?
        get() = locationEngine.getLastLocation()?.let { locationEngine.lastEffectiveSpeed }

    /**
     * Called when the GPS speed state changes.
     */
    fun onSpeedStateChange(isMoving: Boolean) {
        if (!isMoving && coreCoordinator.isAccelMoving()) {
            // Speed SM declared stationary (30s of speed < 1.5 m/s) but accel
            // still reports moving. This could be:
            //   a) Hand tremor while physically still (GPS speed ≈ 0 m/s)
            //   b) Walking slowly (GPS speed 0.3-0.5 m/s, below speed threshold)
            // Only override accel for case (a) — truly near-zero GPS speed.
            // For case (b), trust the accelerometer: the user IS moving.
            //
            // #333: this used to read the raw `Location.speed`, which reports
            // 0.0 on a fix that carries no speed at all, so an unknown speed
            // passed the near-zero test and overruled a positive motion signal
            // on missing data. `lastEffectiveSpeed` is the resolved value — it
            // falls back to a distance/time derivation exactly when the platform
            // reading is absent — and no resolved speed at all now leaves the
            // accelerometer standing rather than silently siding against it.
            val lastSpeed = resolvedSpeed
            when {
                lastSpeed == null ->
                    logger.info("SmartMotionCoordinator: no GPS speed resolved yet — trusting accel, staying continuous.")
                lastSpeed <= TREMOR_SPEED_THRESHOLD -> {
                    logger.info("SmartMotionCoordinator: GPS speed near zero ($lastSpeed m/s) but accel moving — overriding accel to false (hand tremor).")
                    coreCoordinator.onAccelStateChange(false)
                }
                else ->
                    logger.info("SmartMotionCoordinator: GPS speed $lastSpeed m/s is above the $TREMOR_SPEED_THRESHOLD m/s tremor threshold — trusting accel, staying continuous.")
            }
        }
        val action = coreCoordinator.onSpeedStateChange(isMoving)
        logger.debug("SmartMotionCoordinator: onSpeedStateChange -> isMoving=$isMoving, action=$action, isAccelMoving=${coreCoordinator.isAccelMoving()}, isSpeedMoving=${coreCoordinator.isSpeedMoving()}")
        handleAction(action)
    }

    /**
     * Called when the user manually forces the pace via changePace().
     */
    fun onManualPaceChange(isMoving: Boolean) {
        val accelAction = coreCoordinator.onAccelStateChange(isMoving)
        handleAction(accelAction)
        
        val speedAction = coreCoordinator.onSpeedStateChange(isMoving)
        handleAction(speedAction)
    }
    
    /**
     * Synchronize the Rust core mode with the native StateManager on startup or mode change.
     *
     * `stateManager.trackingMode` and the coordinator's `currentMode` measure
     * different things, and conflating them wedged the coordinator (#344):
     *
     * * `trackingMode` is the **session** mode — which start API was called. It
     *   is set to CONTINUOUS by `start()` and stays there for the whole session;
     *   the stationary switch records the pace in `stateManager.isMoving`.
     * * `currentMode` is the **posture** — whether the engine is running
     *   continuous location updates right now or is parked in a stationary
     *   schedule.
     *
     * `start()` sets `isMoving` from `motion.isMoving` (false by default) and
     * then calls this, so mapping CONTINUOUS -> CONTINUOUS wrote CONTINUOUS into
     * a coordinator whose inputs both said stationary. The core's
     * `evaluate_state` emits no action for that pair in either direction: a
     * moving accelerometer sees `currentMode == CONTINUOUS` and returns None, so
     * the session stayed parked — no fixes recorded, nothing to sync — until the
     * process was killed. Reading the posture off `isMoving` keeps the two in
     * step, and the first real accel or speed event produces the wake-up.
     *
     * #409 is the same wedge from the other side: `isMoving` false with a *live*
     * stream parked the coordinator while the engine kept streaming, and
     * `evaluate_state` returns None for that pair too. The posture is now the OR
     * of the committed pace and the engine's actual state — the only reading
     * that survives both directions.
     */
    fun syncCurrentMode() {
        val useGeofences = configManager.getStationaryTrackingMode() ==
            com.ikolvi.tracelet.sdk.model.StationaryTrackingMode.GEOFENCES
        val stationaryMode = if (useGeofences) {
            uniffi.tracelet_core.TrackingMode.STATIONARY_GEOFENCES
        } else {
            uniffi.tracelet_core.TrackingMode.STATIONARY_PERIODIC
        }
        val mode = when (stateManager.trackingMode) {
            TrackingMode.CONTINUOUS ->
                // `isMoving` predicts what start() is *about* to do; `isTracking`
                // reports what the engine is doing *now*. Either alone gets a case
                // wrong, so the posture is the OR of them (#409).
                //
                // `isMoving` alone was the wedge. start() syncs before it touches
                // the engine, so a session resuming stationary while a stream was
                // still live — a ready() takeover after the boot engine handed
                // over — wrote StationaryPeriodic into a coordinator whose engine
                // was streaming. evaluate_state only emits the stop action when it
                // believes the mode is Continuous, so every later stationary
                // decision returned None and the stream ran at the configured
                // interval for the rest of the session, on a device the SDK
                // correctly reported as parked.
                //
                // `isTracking` alone breaks the other direction: this runs before
                // start() starts the engine, so a moving start would sync
                // stationary and then stream — the same wedge, mirrored.
                if (stateManager.isMoving || locationEngine.isTracking) {
                    uniffi.tracelet_core.TrackingMode.CONTINUOUS
                } else {
                    stationaryMode
                }
            TrackingMode.GEOFENCES -> uniffi.tracelet_core.TrackingMode.STATIONARY_GEOFENCES
            TrackingMode.PERIODIC -> uniffi.tracelet_core.TrackingMode.STATIONARY_PERIODIC
            else -> uniffi.tracelet_core.TrackingMode.CONTINUOUS
        }
        coreCoordinator.setCurrentMode(mode)
        coreCoordinator.setUseGeofencesWhenStationary(useGeofences)
    }

    /**
     * Records a tracking-mode switch on the always-on lifecycle channel, naming
     * both inputs of the OR (#334).
     *
     * The coordinator is a logical OR of the accelerometer and the GPS-speed
     * machine, so a downgrade means *both* said stationary. Which one flipped
     * last — and what the other was doing at the time — is the first question
     * asked of any "it stopped tracking while I was moving" report, and at the
     * shipped log levels the answer was not recorded anywhere.
     */
    private fun recordModeSwitch(mode: String) {
        val speedText = resolvedSpeed?.toString() ?: "unknown"
        com.ikolvi.tracelet.sdk.util.TraceletLog.lifecycle(
            "smart-motion: switching to $mode — accelMoving=${coreCoordinator.isAccelMoving()} " +
                "speedMoving=${coreCoordinator.isSpeedMoving()} lastSpeed=${speedText}m/s",
        )
    }

    private fun handleAction(action: uniffi.tracelet_core.CoordinatorAction) {
        when (action) {
            uniffi.tracelet_core.CoordinatorAction.SWITCH_TO_CONTINUOUS -> {
                recordModeSwitch("CONTINUOUS")
                val useForeground = configManager.isForegroundServiceEnabled()
                logger.debug("SmartMotionCoordinator: SWITCH_TO_CONTINUOUS — useForeground=$useForeground")
                if (useForeground) {
                    LocationService.switchToContinuous(locationEngine, stateManager)
                } else {
                    PeriodicLocationWorker.cancel(context)
                    locationEngine.start()
                }
                stateManager.isMoving = true
                logger.debug("SmartMotionCoordinator: SWITCH_TO_CONTINUOUS — calling motionDetector.onManualPaceChange(true)")
                motionDetector.onManualPaceChange(true)
                
                // Dispatch motionchange event
                val locationMap = locationEngine.getLastLocation()?.let {
                    locationEngine.enrichLocation(it, "motionchange")
                } ?: mapOf("is_moving" to true)
                events.sendMotionChange(locationMap)
            }
            uniffi.tracelet_core.CoordinatorAction.SWITCH_TO_STATIONARY_GEOFENCES -> {
                recordModeSwitch("STATIONARY_GEOFENCES")
                val useForeground = configManager.isForegroundServiceEnabled()
                if (useForeground) {
                    LocationService.switchToStationaryGeofences(locationEngine, stateManager, configManager)
                } else {
                    if (configManager.getGeofenceModeHighAccuracy()) {
                        locationEngine.start()
                    } else {
                        locationEngine.stop()
                    }
                }
                stateManager.isMoving = false
                motionDetector.onManualPaceChange(false)
                
                val locationMap = locationEngine.getLastLocation()?.let {
                    locationEngine.enrichLocation(it, "motionchange")
                } ?: mapOf("is_moving" to false)
                events.sendMotionChange(locationMap)
            }
            uniffi.tracelet_core.CoordinatorAction.SWITCH_TO_STATIONARY_PERIODIC -> {
                recordModeSwitch("STATIONARY_PERIODIC")
                val useForeground = configManager.isForegroundServiceEnabled()
                logger.debug("SmartMotionCoordinator: SWITCH_TO_STATIONARY_PERIODIC — useForeground=$useForeground")
                if (useForeground) {
                    LocationService.switchToStationaryPeriodic(locationEngine, configManager, stateManager)
                } else {
                    locationEngine.stop()
                    val lastLoc = locationEngine.getLastLocation()
                    if (lastLoc != null) {
                        stateManager.lastPeriodicLatitude = lastLoc.latitude
                        stateManager.lastPeriodicLongitude = lastLoc.longitude
                        stateManager.lastLocationTime = lastLoc.time
                    }
                    val interval = configManager.getStationaryPeriodicInterval()
                    
                    val useExactAlarms = configManager.getPeriodicUseExactAlarms() || interval < 900
                    if (useExactAlarms) {
                        PeriodicLocationWorker.scheduleOneTime(context)
                        PeriodicLocationWorker.scheduleExactAlarm(context, interval)
                    } else {
                        PeriodicLocationWorker.schedule(context, interval)
                    }
                }
                stateManager.isMoving = false
                logger.debug("SmartMotionCoordinator: SWITCH_TO_STATIONARY_PERIODIC — calling motionDetector.onManualPaceChange(false)")
                motionDetector.onManualPaceChange(false)
                
                val locationMap = locationEngine.getLastLocation()?.let {
                    locationEngine.enrichLocation(it, "motionchange")
                } ?: mapOf("is_moving" to false)
                events.sendMotionChange(locationMap)
            }
            uniffi.tracelet_core.CoordinatorAction.NONE -> {
                // Do nothing
            }
        }
    }

    private companion object {
        /**
         * GPS speed at or below which a *moving* accelerometer is read as hand
         * tremor on a physically still device rather than as real motion.
         */
        const val TREMOR_SPEED_THRESHOLD = 0.15
    }
}
